CORE-16184: SessionClose refactor

emilybowe opened the pull request:
SessionClose refactor as part of CORE-15757/flow-session-refactor 
Now a requireClose boolean value is used. 
If a party sets requireClose to be true it will wait to receive a close (but will not ScheduleClose) and will cleanup.
If requireClose is false then after a configurable amount of time it will ScheduleClose, cleanup, and will not wait for Acks.

LWogan committed:
CORE-15757 some initial changes to setup the feature branch with avro objects to begin development

LWogan committed:
CORE-15757 some initial changes to setup the feature branch with avro objects to begin development

LWogan committed:
CORE-15757 disabled some tests

LWogan committed:
CORE-15757 detekt

LWogan committed:
CORE-15757 disabled test

LWogan committed:
CORE-15757 updated api version

LWogan committed:
CORE-15757 compile error

LWogan committed:
CORE-15757 add require close to the flow session code. Last of initial code changes to feature to allow split of work

LWogan committed:
CORE-15757 disable e2e tests while dev is ongoing

LWogan committed:
CORE-15757 compile error

LWogan committed:
CORE-15757 fix som tests

LWogan committed:
CORE-16184 some comments on where to change code

LWogan committed:
CORE-16184 combined worker e2e test disabled

LWogan committed:
Merge remote-tracking branch 'origin/release/os/5.1' into feature/CORE-15757/flow-session-refactor

LWogan committed:
CORE-15757 update api version

LWogan committed:
Merge remote-tracking branch 'origin/release/os/5.1' into feature/CORE-15757/flow-session-refactor

# Conflicts:
#	gradle.properties

LWogan committed:
CORE-15757 merge from release

emilybowe committed:
changes to SessionCloseProcessorSend and add Utils and Helper for get init-ed identity

emilybowe committed:
kdocs

emilybowe committed:
CloseSessionsRequestHandler WaitingFor complete and postProcess in progress

emilybowe committed:
working on CloseSessionsRequestHandler

emilybowe committed:
use value parameters and change string to bool

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/CloseSessionsRequestHandler.kt:
```diff
@@ -30,7 +34,7 @@ class CloseSessionsRequestHandler @Activate constructor(
     override fun getUpdatedWaitingFor(context: FlowEventContext<Any>, request: FlowIORequest.CloseSessions): WaitingFor {
         val sessionsToClose = try {
-            getSessionsToClose(context.checkpoint, request)
+            getSessionsToClose(context.checkpoint, request, initiatingIdentity = true)
```
You don't know at this point if it was the initiator or initiated who called close here. We can suspend for both cases.

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/CloseSessionsRequestHandler.kt:
```diff
@@ -67,9 +88,63 @@ class CloseSessionsRequestHandler @Activate constructor(
+         && filter out sessions in status ERROR or CLOSED
+         */
+
+        val sessions = request.sessions.toList()
```
the kdoc comments i added above here is logic you need to apply to each sessionId in this list.

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/CloseSessionsRequestHandler.kt:
```diff
@@ -67,9 +88,63 @@ class CloseSessionsRequestHandler @Activate constructor(
+        val statusToFilterOut = setOf(SessionStateType.ERROR, SessionStateType.CLOSED)
+
+        val sessionsInErrorOrClosed = flowSessionManager.getSessionsWithStatuses(checkpoint, sessions, statusToFilterOut)
+        sessions - sessionsInErrorOrClosed.map { it.sessionId }
```
This operation isn't assigned to any field so it has no affect. Needs to be something like:
val filteredSessionsByStatus = sessions - sessionsInErrorOrClosed.map { it.sessionId }

emilybowe replied:
I added the -= if you're happy with that:
e.g.         `sessions -=  sessionsInErrorOrClosed.map { it.sessionId }`

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/CloseSessionsRequestHandler.kt:
```diff
@@ -67,9 +88,63 @@ class CloseSessionsRequestHandler @Activate constructor(
+        sessions - sessionsInErrorOrClosed.map { it.sessionId }
+
+        val requiredCloseSessions = flowSessionManager.getSessionsByRequireClose(checkpoint, sessions, requireClose = true)
+        sessions - requiredCloseSessions.map { it.sessionId }
```
same problem here.

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/CloseSessionsRequestHandler.kt:
```diff
@@ -67,9 +88,63 @@ class CloseSessionsRequestHandler @Activate constructor(
+        val requiredCloseSessions = flowSessionManager.getSessionsByRequireClose(checkpoint, sessions, requireClose = true)
+        sessions - requiredCloseSessions.map { it.sessionId }
+
+        return if(initiatingIdentity) {
```
Same comment as earlier. We need to work out if we are the initiating identity per sessionId in this handler. In reality due to how the flow engin/session APIs are written we are always one or the other for all sessions in the list, but we shouldn't code based on that here.

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/sessions/FlowSessionManager.kt:
```diff
@@ -237,4 +237,35 @@ interface FlowSessionManager {
+    fun getSessionsByRequireClose(
+        checkpoint: FlowCheckpoint,
+        sessionIds: List<String>,
+        requireClose: Boolean
```
requireClose will be set per session, so you can't have a boolean on this API.

emilybowe replied:
The boolean is there so I know if when I'm filtering the list I'm interested in returning for requireClose = true or requireClose = false
Is that use case ok?

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/sessions/FlowSessionManager.kt:
```diff
@@ -237,4 +237,35 @@ interface FlowSessionManager {
+     *
+     * @throws []FlowSessionStateException] If a session does not exist within the flow's [FlowCheckpoint].
+     */
+    fun getSessionsWithStatus(
```
Theres already a getSessionsWithStatus in this class

LWogan commented on libs/flows/session-manager-impl/src/main/kotlin/net/corda/session/manager/impl/processor/SessionCloseProcessorSend.kt:
```diff
@@ -103,28 +117,29 @@ class SessionCloseProcessorSend(
+    ) : SessionState {
+        val requireClose = sessionState.requireClose
+        var status = sessionState.status
+        return if (isInitiatedIdentity(sessionEvent) && status !== SessionStateType.CLOSED) {
```
this class looks correct!

LWogan replied:
I think if the status was ERROR this would also be a condition where we shouldn't try send an close.
Change the second condition to something like 

``status !in listOf(SessionStateType.ERROR, SessionStateType.CLOSED)``

LWogan commented on libs/flows/session-manager-impl/src/main/kotlin/net/corda/session/manager/impl/processor/SessionCloseProcessorSend.kt:
```diff
@@ -103,28 +117,29 @@ class SessionCloseProcessorSend(
-                    "$currentState. Current SessionState: $sessionState."
+        } else {
+            val errorMessage =
+                "Tried to send SessionClose on key $key and sessionId $sessionId, session status is " +
```
need to update this error to say something about the fact its trying to send a close when requireClose is set to false. It shouldn't ever happen in reality.

LWogan reviewed (commented):
I think a few of my comments are in classes that are still in progress but wanted to make sure we course correct early.

emilybowe committed:
SessionClose PR comment change msg

emilybowe committed:
add -= operator, address PR comments

emilybowe committed:
changes to getSessionsToClose

emilybowe committed:
updateStatus api

emilybowe committed:
adding getSessionsBySessionSide

emilybowe committed:
changes to getInitiatingAndInitiatedSessions

emilybowe committed:
refactor and make getSessionsToCloseForWaitingFor, rm lines in SessionCloseProcessorReceive

emilybowe committed:
refactoring and writing postProcess

emilybowe committed:
work on postProcess

emilybowe committed:
changes to updateStatus

emilybowe committed:
tidy up

emilybowe committed:
fix detekt

LWogan commented on libs/flows/session-manager-impl/src/main/kotlin/net/corda/session/manager/impl/processor/SessionCloseProcessorSend.kt:
```diff
@@ -32,6 +33,19 @@ class SessionCloseProcessorSend(
         private val logger = LoggerFactory.getLogger(this::class.java.enclosingClass)
     }
 
+    /**
```
remove these pseudo code comments throughout the code if they're implemented

LWogan commented on libs/flows/session-manager-impl/src/main/kotlin/net/corda/session/manager/impl/processor/SessionCloseProcessorSend.kt:
```diff
@@ -103,28 +117,29 @@ class SessionCloseProcessorSend(
-                    "$currentState. Current SessionState: $sessionState."
+        } else {
+            val errorMessage =
+                "Tried to send SessionClose when requireClose is set to false. Key: $key sessionId: $sessionId, session status is " +
```
this error message isn't correct based on the If condition.  The issue would be an initiating party trying to send a close message which is not allowed in the protocol or the status is not correct for sending a close

LWogan commented on libs/flows/session-manager-impl/src/main/kotlin/net/corda/session/manager/impl/processor/SessionCloseProcessorSend.kt:
```diff
@@ -103,28 +117,29 @@ class SessionCloseProcessorSend(
+                                "${sessionEvent.sequenceNum}, $sessionId"
+                    }
+                    status = SessionStateType.CLOSED
+                    sendEventsState.lastProcessedSequenceNum = nextSeqNum
```
We also need to set session event sequence number

sessionEvent.sequenceNum = nextSeqNum

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/CloseSessionsRequestHandler.kt:
```diff
@@ -49,10 +50,48 @@ class CloseSessionsRequestHandler @Activate constructor(
+                }
+            }
+
+            //TODO - @Lorcan Can I delete this?
```
no

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/CloseSessionsRequestHandler.kt:
```diff
@@ -49,10 +50,48 @@ class CloseSessionsRequestHandler @Activate constructor(
+            }
+
+            /**
+             *  if i am initiating party i.e i dont end with INITIATED_SESSION_ID_SUFFIX
```
remove this comment

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/CloseSessionsRequestHandler.kt:
```diff
@@ -67,9 +106,50 @@ class CloseSessionsRequestHandler @Activate constructor(
+    private fun getSessionsToClose(request: FlowIORequest.CloseSessions): List<String> {
+        /**
+         * else if i am initiating party i.e i dont end with INITIATED_SESSION_ID_SUFFIX
+        if requireClose == false
```
remove this

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/sessions/impl/Constants.kt:
```diff
@@ -0,0 +1,4 @@
+package net.corda.flow.pipeline.sessions.impl
+
+const val INITIATED = "initiated"
```
are these being used?

LWogan reviewed (commented):
Some initial comments

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/CloseSessionsRequestHandler.kt:
```diff
@@ -49,10 +50,48 @@ class CloseSessionsRequestHandler @Activate constructor(
+                    flowSessionManager.updateStatus(checkpoint, requireCloseFalse, SessionStateType.CLOSED)
+                }
+
+                if (requireCloseTrue.isNotEmpty()) {
```
If the initiating party calls close() and the they have already received a close from the other side, then the current status will be CLOSING. In that case we would want to update those session states to CLOSED instead of CLOSING. 

The logic in this class is starting to get a bit confusing with multiple calls to the flow session manager.

Given this logic needs to be replicated to SubFlowFinishedRequestHandler, i think moving the logic from line 78 -> 91 to its own service might help make the code more maintainable

emilybowe replied:
handled here: closeSessionsAlreadyInClosing(checkpoint, initiatingSessions)
and I've split out the logic

emilybowe committed:
CORE-15069 move flow to components (#4164)

Move two flow modules from libs to components. This will be used in future work to make use of a membership service in components to check clusters are local. Moving or copying this service was considered also but the move is also suggested for better design as these modules are closer to components than libs.

emilybowe committed:
CORE-15070: use service to check if cluster is local (#4201)

This PR adds a RecordFactory and slightly redesigns Init, Error and Event Executors. These executors now check if Outbound events are destined for the local cluster via the record factory and selects which topic to publish to. The RecordFactory uses a lookup service from membership to do this.

LWogan committed:
Merge remote-tracking branch 'origin/release/os/5.1' into feature/CORE-15757/flow-session-refactor

# Conflicts:
#	gradle.properties

LWogan committed:
CORE-15757 merge from release branch

emilybowe committed:
fixing NPE in unit tests for CloseSessionsRequestHandlerTest for new methods on flowSessionsManager

emilybowe committed:
extact to a new service

emilybowe committed:
changing subFlowFinished

emilybowe committed:
refactor to CloseSessionsRequestHandler

emilybowe committed:
more changes to subFlowFinishedHandler

emilybowe committed:
error messages and update sessionEvent seq num and refactors

emilybowe committed:
detekt fix

emilybowe committed:
add service to test

emilybowe committed:
fix detekt

emilybowe committed:
fix detekt

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/service/CloseSessionService.kt:
```diff
@@ -0,0 +1,74 @@
+                flowSessionManager.updateStatus(checkpoint, requireCloseFalse, SessionStateType.CLOSED)
+            }
+
+            if (requireCloseTrue.isNotEmpty()) {
```
in closeSessionsAlreadyInClosing you set sessions that were in CLOSING to CLOSED, but then here you would reset them back to CLOSING. If the status is ERROR/CLOSED you dont want to update the session state as its already  terminated. 


Try find the cleanest way possible to handle the possible scenarios for initiating flows are
- status == ERROR/CLOSED-> do nothing, sessions are already terminated
- requireClose == true && status == CREATED/CONFIRMED -> go to CLOSING
- requireClose == true && status == CLOSING -> go to CLOSED
- requireClose == false -> go to CLOSED.

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/service/CloseSessionService.kt:
```diff
@@ -0,0 +1,74 @@
+import java.time.Instant
+
+@Component(service = [CloseSessionService::class])
+class CloseSessionService @Activate constructor(
```
no kdocs on the class or any of the methods

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/service/CloseSessionService.kt:
```diff
@@ -0,0 +1,74 @@
+        return requireCloseSessions.first
+    }
+
+    fun getSessionsForPostProcess(sessionsToClose: List<String>, checkpoint: FlowCheckpoint) {
```
this is called get but doesn't return anything. it should be named something like handleCloseForSessions or event closeSessions. (im not the best with names)

LWogan reviewed (commented):
Couple more comments

emilybowe committed:
PR comments - kdocs, name change, change filter

emilybowe committed:
fixing CloseSessionsRequestHandlerTest

emilybowe committed:
SubFlowFinishedRequestHandlerTest fixes

LWogan committed:
CORE-16184 Pair programming refactor

emilybowe committed:
fix CloseSessionsRequestHandlerTest after refactoring

emilybowe committed:
fix SubFlowFinishedRequestHandlerTest after refactoring

LWogan committed:
CORE-16174 session init refactor, removal of heartbeating and resends (#4483)

SessionInit nullable field added to SessionData object. Populated for each data message sent to the initiated party until they respond with a data message. Initiator sessionState status will remain in CREATED until they receive a data message back (CONFIRMED)
We cannot be sure which SessionData will arrive first to the counterparty due to no order guarantees in the p2p layer. SessionInit object contains the information needed to start a flow on the other side, as well as the user and platform context data. This allows the checkpoint to be created regardless of which data message arrives first.

Removed restriction from initiating flow whereby they could not proceed past the first send without receiving something back.

contextSessionProperties is piggybacked onto existing data messages when it is not explicitly requested. It is moved to the SessionEvent object rather than the SessionConfirm object.

SessionInit & Confirm exchange is kept for when session.getCounterpartyInfo is called before any messages are exchanged.

Pure Acks removed

Heartbeating is removed

Session timeout logic left in place. Additional work required to integrate with the scheduler

Message resends logic removed.

A simple close mechanism implemented pending future work

LWogan committed:
CORE-16174 update api version

LWogan committed:
Merge remote-tracking branch 'origin/feature/CORE-15757/flow-session-refactor' into feature/CORE-15757/flow-session-refactor

# Conflicts:
#	gradle.properties

emilybowe committed:
writing CloseSessionService tests

LWogan committed:
Merge remote-tracking branch 'origin/release/os/5.1' into feature/CORE-15757/flow-session-refactor

# Conflicts:
#	gradle.properties

emilybowe committed:
CloseSessionServiceTest and some kdoc edits

emilybowe committed:
fixed updateStatus so it will update status per parameter and not change to CLOSED regardless, writing tests for flowManagerImpl

emilybowe committed:
writing tests for flowManagerImpl, added requireClose tests

emilybowe committed:
add test for getInitiatingAndInitiatedSessions

emilybowe committed:
add new test for close when session state is CLOSED

emilybowe committed:
Merge branch 'feature/CORE-15757/flow-session-refactor' into CORE-16184/session-close-refactor

emilybowe committed:
add default of false for requireClose in buildSessionState

emilybowe committed:
enable CloseSessionsRequestHandlerTest and SubFlowFinishedRequestHandlerTest

emilybowe committed:
leftover merge conflict

emilybowe committed:
add refactor of subFlowFinishedHandler back in

emilybowe committed:
CORE-16762 fixing tests disabled by flow session refactor now enabled - CloseSessionsRequestHandlerTest and SubFlowFinishedRequestHandlerTest now passing

emilybowe committed:
fix detekt

emilybowe committed:
fix 'Send close when status is CONFIRMED' test in SessionCloseProcessorSend

emilybowe committed:
fixed 'Send close when status is already CLOSING due to close received()' and changed to 'If state is CLOSING or CLOSED, no update required' as protocol has changed

emilybowe committed:
fix 'Send close when in state confirmed'

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/helper/Utils.kt:
```diff
@@ -0,0 +1,15 @@
+ * @param sessionEvent input session event to get identity info from
+ * @return [Boolean] is this event sent by the Initiating identity
+ */
+fun isInitiatingIdentity(sessionId: String): Boolean {
```
Theres a common helper method in :libs:flows:flow-utils which you can use instead of adding this new method. Beware it is the inverse of your function, returning true for initiated, not initiating. so youll need to account for that

fun isInitiatedParty(sessionEvent: SessionEvent) : Boolean {
    return sessionEvent.sessionId.contains(INITIATED_SESSION_ID_SUFFIX)
}

emilybowe replied:
do you object to this:

```
package net.corda.flow.utils

const val INITIATED_SESSION_ID_SUFFIX = "-INITIATED"

fun isInitiatedIdentity(sessionId: String) : Boolean {
    return sessionId.contains(INITIATED_SESSION_ID_SUFFIX)
}

fun isInitiatingIdentity(sessionId: String) : Boolean {
    return !sessionId.contains(INITIATED_SESSION_ID_SUFFIX)
}
```

I think it's easier to read isInitiated / isInitiating when I was writing this but I'll change it if you feel it's better to remove the method. I did change all the usages so they are all coming from that one utils file.

LWogan replied:
yeah thats ok with me.

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/application/sessions/impl/FlowSessionImpl.kt:
```diff
@@ -116,8 +117,12 @@ class FlowSessionImpl(
-        log.trace { "Closing session: $sourceSessionId" }
+        val flowCheckpoint = flowFiberService.getExecutingFiber().getExecutionContext().flowCheckpoint
+        val sessionState = flowCheckpoint.getSessionState(sourceSessionId)
+        if(checkNotNull(sessionState!!.status) != SessionStateType.CLOSED) {
```
If session state was null here you'd throw a null pointer exception so the checkNotNull isn't doing anything functional. 

It should be impossible for the session state to be null at this point. If it is null and we suspend the close request handler will detect it and throw an appropriate exception. So im ok with checking status against closed where status can be null i.e

```
val sessionState = flowFiberService.getExecutingFiber().getExecutionContext().flowCheckpoint.getSessionState(sourceSessionId)
        if(sessionState?.status != SessionStateType.CLOSED) {
            fiber.suspend(FlowIORequest.CloseSessions(setOf(sourceSessionId)))
            log.trace { "Closing session: $sourceSessionId" }
        }

LWogan commented on components/flow/flow-service/src/test/kotlin/net/corda/flow/pipeline/sessions/impl/FlowSessionManagerImplTest.kt:
```diff
@@ -52,6 +53,7 @@ import java.nio.ByteBuffer
 import java.time.Instant
 import java.util.stream.Stream
 
+@Disabled //todo - core-15757
```
you need to enable these tests

LWogan commented on libs/flows/session-manager-impl/src/main/kotlin/net/corda/session/manager/impl/processor/SessionCloseProcessorSend.kt:
```diff
@@ -85,37 +80,44 @@ class SessionCloseProcessorSend(
+        } else {
+            val status = sessionState.status
+            val errorMessage: String = if (status in listOf(SessionStateType.ERROR, SessionStateType.CLOSED)) {
+                "Tried to send SessionClose when status is not correct for sending a close. " +
```
I think you can write a single error message that conveys all the necessary information required for for the log without the conditionals.

LWogan commented on libs/flows/session-manager-impl/src/main/kotlin/net/corda/session/manager/impl/processor/helper/SessionProcessorHelper.kt:
```diff
@@ -140,4 +141,13 @@ fun setErrorState(
+ * @param sessionEvent input session event to get identity info from
+ * @return [Boolean] is this event sent by the Initiated identity
+ */
+fun isInitiatedIdentity(sessionEvent: SessionEvent): Boolean {
```
delete and reuse the new flow-utils method

LWogan reviewed (commented):
Couple more comments. Address all the comments before you merge from release again as there will be a headache merge to do when the flow session refactor pr is merged today. Having your branch ready to go makes it marginally easier for you.

emilybowe committed:
change new api CloseSessionService parameter order so it is consistent with FlowSessionManager order

emilybowe committed:
rm usage of helper files identity check and check from libs-flow-utils method instead

emilybowe committed:
rm checkNotNull

emilybowe committed:
enable flow session manager impl tests and change log message as per pr comments

emilybowe committed:
Merge remote-tracking branch 'origin/release/os/5.1' into CORE-16184/session-close-refactor

# Conflicts:
#	components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/context/FlowIoRequestBuilders.kt
#	components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/CloseSessionsAcceptanceTest.kt
#	components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/FlowFailedAcceptanceTest.kt
#	components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/FlowFinishedAcceptanceTest.kt
#	components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/InitiateFlowAcceptanceTest.kt
#	components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/ReceiveAcceptanceTest.kt
#	components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/SendAcceptanceTest.kt
#	components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/SendAndReceiveAcceptanceTest.kt
#	components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/SubFlowFailedAcceptanceTest.kt
#	components/flow/flow-service/src/integrationTest/kotlin/net/corda/flow/testing/tests/SubFlowFinishedAcceptanceTest.kt
#	components/flow/flow-service/src/main/kotlin/net/corda/flow/application/services/impl/FlowMessagingImpl.kt
#	components/flow/flow-service/src/main/kotlin/net/corda/flow/application/sessions/impl/FlowSessionImpl.kt
#	components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/events/SessionEventHandler.kt
#	components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/SubFlowFinishedRequestHandler.kt
#	components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/CloseSessionsRequestHandler.kt
#	components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/SendAndReceiveRequestHandler.kt
#	components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/SendRequestHandler.kt
#	components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/waiting/SessionInitWaitingForHandler.kt
#	components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/runner/impl/FlowRunnerImpl.kt
#	components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/sessions/impl/FlowSessionManagerImpl.kt
#	components/flow/flow-service/src/test/kotlin/net/corda/flow/application/sessions/FlowSessionImplTest.kt
#	components/flow/flow-service/src/test/kotlin/net/corda/flow/pipeline/handlers/events/SessionEventHandlerTest.kt
#	components/flow/flow-service/src/test/kotlin/net/corda/flow/pipeline/handlers/requests/SubFlowFinishedRequestHandlerTest.kt
#	components/flow/flow-service/src/test/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/CloseSessionsRequestHandlerTest.kt
#	components/flow/flow-service/src/test/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/SendRequestHandlerTest.kt
#	components/flow/flow-service/src/test/kotlin/net/corda/flow/pipeline/handlers/waiting/sessions/SessionConfirmationWaitingForHandlerTest.kt
#	components/flow/flow-service/src/test/kotlin/net/corda/flow/pipeline/handlers/waiting/sessions/SessionInitWaitingForHandlerTest.kt
#	components/flow/flow-service/src/test/kotlin/net/corda/flow/pipeline/impl/FlowMDCServiceImplTest.kt
#	components/flow/flow-service/src/test/kotlin/net/corda/flow/pipeline/runner/FlowRunnerImplTest.kt
#	gradle.properties
#	libs/flows/flow-utils/src/main/kotlin/net/corda/flow/utils/FlowUtils.kt
#	libs/flows/session-manager-impl/src/main/kotlin/net/corda/session/manager/impl/SessionManagerImpl.kt
#	libs/flows/session-manager-impl/src/main/kotlin/net/corda/session/manager/impl/processor/SessionCloseProcessorReceive.kt
#	libs/flows/session-manager-impl/src/main/kotlin/net/corda/session/manager/impl/processor/SessionCloseProcessorSend.kt
#	libs/flows/session-manager-impl/src/main/kotlin/net/corda/session/manager/impl/processor/SessionDataProcessorSend.kt
#	libs/flows/session-manager-impl/src/test/kotlin/net/corda/session/manager/impl/processor/SessionCloseProcessorSendTest.kt
#	processors/flow-processor/src/main/kotlin/net/corda/processors/flow/internal/FlowProcessorImpl.kt
#	simulator/runtime/src/main/kotlin/net/corda/simulator/runtime/messaging/ConcurrentFlowMessaging.kt

emilybowe committed:
fix merge conflicts

emilybowe committed:
merge conflict fix

emilybowe committed:
fix merge conflicts

emilybowe committed:
rm check for put session states as this is in close session service

emilybowe committed:
Merge remote-tracking branch 'origin/release/os/5.1' into CORE-16184/session-close-refactor

# Conflicts:
#	components/flow/flow-service/src/test/kotlin/net/corda/flow/RequestHandlerTestContext.kt

LWogan committed:
CORE-16184 add session close event to received events

LWogan committed:
CORE-16184 fix acceptance tests to handle when close is called. Flows will suspend here and wait for close message before proceeding as requireTrue defaults to true

LWogan committed:
Merge remote-tracking branch 'origin/release/os/5.1' into CORE-16184/session-close-refactor

LWogan committed:
CORE-16184 additional acceptance tests for close and sub flow finished handlers

LWogan committed:
CORE-16184 update session integration test

LWogan committed:
CORE-16184 fix unit test

LWogan commented on components/flow/flow-service/src/main/kotlin/net/corda/flow/pipeline/handlers/requests/sessions/service/CloseSessionService.kt:
```diff
@@ -0,0 +1,94 @@
+     * @param sessions List of sessions IDs.
+     * @return List of sessions that are not CLOSED or ERROR
+     */
+    fun handleCloseForSessions(checkpoint: FlowCheckpoint, sessions: List<String>): List<String> {
```
This method does not need to return anything. it just updates the session states.

LWogan reviewed (commented):
One more comment and then i think we're good to merge!

emilybowe committed:
tidy ups

emilybowe committed:
rm return that's not needed

emilybowe committed:
rm return in tests to fix compile

emilybowe committed:
tidy up and enable tests

emilybowe committed:
Merge remote-tracking branch 'origin/release/os/5.1' into CORE-16184/session-close-refactor

emilybowe committed:
fix test that calls getSessionWithStatus one less time after change

LWogan reviewed (approved):
LGTM
